src/bvh/bvh.rs

```diff
@@ -0,0 +1,444 @@
+use ray::Ray;
+use std::f32;
+use std::iter::repeat;
+
+/// Enum which describes the union type of a node in a [`BVH`].
+/// This structure does not allow for storing a root node's [`AABB`]. Therefore rays
+/// which do not hit the root [`AABB`] perform two [`AABB`] tests (left/right) instead of one.
+/// On the other hand this structure decreases the total number of indirections when traversing
+/// the BVH. Only those nodes are accessed, which are definetely hit.
+///
+/// [`AABB`]: ../aabb/struct.AABB.html
+/// [`BVH`]: struct.BVH.html
+///
+#[derive(Debug, Copy, Clone)]
+pub enum BVHNode {
+    /// Leaf node.
+    Leaf {
+        /// The node's parent.
+        parent: usize,
+
+        /// The node's depth
+        depth: u32,
+
+        /// The shape contained in this leaf.
+        shape: usize,
+    },
+    /// Inner node.
+    Node {
+        /// The node's parent.
+        parent: usize,
+
+        /// The node's depth
+        depth: u32,
+
+        /// The convex hull of the shapes' `AABB`s in child_l.
+        child_l_aabb: AABB,
+
+        /// Left subtree.
+        child_l: usize,
+
+        /// The convex hull of the shapes' `AABB`s in child_r.
+        child_r_aabb: AABB,
+
+        /// Right subtree.
+        child_r: usize,
+    },
+    /// Dummy node for later replacement
+    Dummy,
+}
+
+impl BVHNode {
+    /// Builds a [`BVHNode`] recursively using SAH partitioning.
+    /// Returns the index of the node in the nodes vector.
+    ///
+    /// [`BVHNode`]: enum.BVHNode.html
+    ///
+    pub fn build<T: BHShape>(shapes: &mut [T],
+                             indices: Vec<usize>,
+                             nodes: &mut Vec<BVHNode>,
+                             parent: usize,
+                             depth: u32)
+                             -> usize {
+        // Helper function to accumulate the AABB joint and the centroids AABB
+        fn grow_convex_hull(convex_hull: (AABB, AABB), shape_aabb: &AABB) -> (AABB, AABB) {
+            let center = &shape_aabb.center();
+            let convex_hull_aabbs = &convex_hull.0;
+            let convex_hull_centroids = &convex_hull.1;
+            (convex_hull_aabbs.join(shape_aabb), convex_hull_centroids.grow(center))
+        }
+
+        let mut convex_hull = Default::default();
+        for index in &indices {
+            convex_hull = grow_convex_hull(convex_hull, &shapes[*index].aabb());
+        }
+        let (aabb_bounds, centroid_bounds) = convex_hull;
+
+        // If there is only one element left, don't split anymore
+        if indices.len() == 1 {
+            let shape_index = indices[0];
+            nodes.push(BVHNode::Leaf {
+                parent: parent,
+                depth: depth,
+                shape: shape_index,
+            });
+            let node_index = nodes.len() - 1;
+            // Let the shape know the index of the node that represents it
+            shapes[shape_index].set_bh_node_index(node_index);
```

Comment:
As now discussed, this isn't very pretty, but other solutions are at least as ugly.
In any case, the system outside of the BVH needs to know the tree node indices in order to be able to tell the optimize function which nodes have changed - and thus are potential rotation candidates.